smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt

```diff
@@ -0,0 +1,205 @@
+    "This field is deprecated and no longer recommended for use."
+
+private const val CLASS_NAME = "EndpointParameters"
+
+/**
+ * Renders the struct of parameters to be passed to the endpoint provider for resolution.
+ */
+class EndpointParametersGenerator(
+    private val writer: KotlinWriter,
+    rules: EndpointRuleset,
+) {
+    private val params: List<KotlinEndpointParameter> = rules.parameters.toList().map {
+        KotlinEndpointParameter(
+            it.name.toKotlin(),
+            it.type.toSymbol(),
+            it.isRequired,
+            it.defaultValue.getOrNull()?.toKotlinLiteral() ?: "null",
+            it.documentation.getOrNull(),
+            it.deprecated.getOrNull(),
+        )
+    }
+
+    fun render() {
+        renderDocumentation()
+        writer.withBlock("public class #L private constructor(builder: Builder) {", "}", CLASS_NAME) {
+            renderFields()
+            renderCompanionObject()
+            write("")
+            if (params.any(KotlinEndpointParameter::isRequired)) {
+                renderInit()
+                write("")
+            }
+            renderEquals()
+            write("")
+            renderHashCode()
+            write("")
+            renderToString()
+            write("")
+            renderCopy()
+            write("")
+            renderBuilder()
+        }
+    }
+
+    private fun renderFields() {
+        params.forEach {
+            writer.ensureSuppressDeprecation(it)
+            it.renderDeclaration(writer, "builder.${it.name}")
+            writer.write("")
+        }
+    }
+
+    private fun renderDocumentation() {
+        writer.dokka {
+            write("The set of values necessary for endpoint resolution.")
+        }
+    }
+
+    private fun renderCompanionObject() {
+        writer.withBlock("public companion object {", "}") {
+            write("public operator fun invoke(block: Builder.() -> Unit): #L = Builder().apply(block).build()", CLASS_NAME)
+        }
+    }
+
+    private fun renderInit() {
+        writer.withBlock("init {", "}") {
+            params.forEach {
+                if (it.isRequired) {
+                    write("""requireNotNull(#1L) { "endpoint provider parameter #1L is required" }""", it.name)
+                }
+            }
+        }
+    }
+
+    private fun renderEquals() {
+        writer.withBlock("public override fun equals(other: Any?): Boolean {", "}") {
+            write("if (this === other) return true")
+            write("if (other !is #L) return false", CLASS_NAME)
+            params.forEach {
+                ensureSuppressDeprecation(it)
+                write("if (this.#1L != other.#1L) return false", it.name)
+            }
+            write("return true")
+        }
+    }
+
+    private fun renderHashCode() {
+        writer.withBlock("public override fun hashCode(): Int {", "}") {
+            if (params.isEmpty()) {
+                write("return 0")
+                return@withBlock
+            }
```

Comment:
**Question**: In `StructureGenerator`, we use `this::class.hashCode()` as the zero-param value which seems like it would provide better hash bucketing in situations where multiple classes are mixed. Are there drawbacks to doing the same here?

Reply:
Yeah unless there is good reason I wouldn't deviate